File: app/src/com/ternaryop/photoshelf/adapter/PhotoShelfPost.java
package com.ternaryop.photoshelf.adapter;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import com.ternaryop.photoshelf.DraftPostHelper;
import com.ternaryop.tumblr.TumblrPhotoPost;

public class PhotoShelfPost extends TumblrPhotoPost {
    /**
	 * 
	 */
	private static final long serialVersionUID = -6670033021694674250L;
	private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("dd/MM/yyyy", Locale.US);
	private long lastPublishedTimestamp;
	private int groupId;

	public enum ScheduleTime {
		POST_PUBLISH_NEVER,
		POST_PUBLISH_FUTURE,
		POST_PUBLISH_PAST
	};
	
	public PhotoShelfPost(TumblrPhotoPost photoPost, long lastPublishedTimestamp) {
		super(photoPost);
		this.lastPublishedTimestamp = lastPublishedTimestamp;
	}

	/**
	 * The last published time can be in the future if the post is scheduled
	 * @return
	 */
	public long getLastPublishedTimestamp() {
		return lastPublishedTimestamp;
	}

	public void setLastPublishedTimestamp(long lastPublishedTimestamp) {
		this.lastPublishedTimestamp = lastPublishedTimestamp;
	}
	
	public ScheduleTime getScheduleTimeType() {
		if (lastPublishedTimestamp == Long.MAX_VALUE) {
			return ScheduleTime.POST_PUBLISH_NEVER;
		} else if (lastPublishedTimestamp > System.currentTimeMillis()) {
			return ScheduleTime.POST_PUBLISH_FUTURE;
		} else {
			return ScheduleTime.POST_PUBLISH_PAST;
		}
	}
	
	public String getLastPublishedTimestampAsString() {
		long tt = getScheduledPublishTime() > 0 ? getScheduledPublishTime() * 1000 : lastPublishedTimestamp;
		long days = DraftPostHelper.daysSinceTimestamp(tt);
		String daysString = DraftPostHelper.formatPublishDaysAgo(tt);
		if (days != Long.MAX_VALUE && days > 0) {
			String string = DATE_FORMAT.format(new Date(tt));
        	daysString += " (" + string + ")";
		}
		return daysString;
	}

	public int getGroupId() {
		return groupId;
	}

	public void setGroupId(int groupId) {
		this.groupId = groupId;
	}
	
	/**
	 * Protect against IndexOutOfBoundsException returning an empty string
	 * @return
	 */
	public String getFirstTag() {
		return getTags().size() > 0 ? getTags().get(0) : "";
	}
}

File: app/src/main/java/com/ternaryop/photoshelf/adapter/PhotoShelfPost.kt
package com.ternaryop.photoshelf.adapter

import android.text.format.DateUtils.SECOND_IN_MILLIS
import com.ternaryop.tumblr.TumblrPhotoPost
import com.ternaryop.utils.DateTimeUtils

/**
 * The last published time can be in the future if the post is scheduled
 */
class PhotoShelfPost(photoPost: TumblrPhotoPost, var lastPublishedTimestamp: Long) : TumblrPhotoPost(photoPost) {
    var groupId: Int = 0

    val scheduleTimeType: ScheduleTime
        get() = if (lastPublishedTimestamp == java.lang.Long.MAX_VALUE) {
            ScheduleTime.POST_PUBLISH_NEVER
        } else if (lastPublishedTimestamp > System.currentTimeMillis()) {
            ScheduleTime.POST_PUBLISH_FUTURE
        } else {
            ScheduleTime.POST_PUBLISH_PAST
        }

    val lastPublishedTimestampAsString: String
        get() {
            val tt = if (scheduledPublishTime > 0) scheduledPublishTime * SECOND_IN_MILLIS else lastPublishedTimestamp
            return DateTimeUtils.formatPublishDaysAgo(tt, DateTimeUtils.APPEND_DATE_FOR_PAST_AND_PRESENT)
        }

    /**
     * Protect against IndexOutOfBoundsException returning an empty string
     * @return the first tag or an empty string
     */
    val firstTag: String
        get() = if (tags.isEmpty()) "" else tags[0]

    enum class ScheduleTime {
        POST_PUBLISH_NEVER,
        POST_PUBLISH_FUTURE,
        POST_PUBLISH_PAST
    }

    companion object {
        private const val serialVersionUID = -6670033021694674250L
    }
}
